### PR TITLE
documentElement.focus() should move focus to the viewport instead of doing nothing

### PR DESCRIPTION
**Pull request by Ahmad-S792**
#### 569bafdafba6da329c22c4ef9437b1a95832b1b1
<pre>
documentElement.focus() should move focus to the viewport instead of doing nothing
<a href="https://bugs.webkit.org/show_bug.cgi?id=322325">https://bugs.webkit.org/show_bug.cgi?id=322325</a>
<a href="https://rdar.apple.com/185567540">rdar://185567540</a>

Reviewed by Ryosuke Niwa.

Element::focus() returns early for an element that is not programmatically
focusable. That is wrong for the document element, because the focusing steps
first substitute the focusable area:

    If new focus target is not a focusable area, then set new focus target to the
    result of getting the focusable area for new focus target, given focus trigger
    if it was passed.

and getting the focusable area has a clause for it:

    If focus target is the document element of its Document

    If focus target is the document element of its Document
        Return the Document&apos;s viewport.

<a href="https://html.spec.whatwg.org/multipage/interaction.html#get-the-focusable-area">https://html.spec.whatwg.org/multipage/interaction.html#get-the-focusable-area</a>

The result is non-null, so the focusing steps proceed and focus moves to the
viewport. The spec makes this explicit in the note on blur(): &quot;if you want to
focus the viewport, call the focus() method on the Document&apos;s document element.&quot;
We instead left the previously focused element focused.

The &quot;element with one or more scrollable regions&quot; clause of getting the focusable
area is still not implemented; only the document element case is handled here.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-closedby-corner-cases-expected.txt: Progression
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-closedby-corner-cases-expected.txt: Removed.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):

Canonical link: <a href="https://commits.webkit.org/319730@main">https://commits.webkit.org/319730@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/33a07c61f852a2d30cd70b7762f809a87a5e122c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/182595 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/56527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/50333 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/192819 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/146892 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ad04e897-0fb6-4440-9fa1-129a8bb81c2b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/184457 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/57858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/56260 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/142499 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/98954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/20a0687b-b880-47c8-9702-2605c1a890ae) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/185550 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/46220 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/163260 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/122933 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bc878d14-1a41-4408-a8a8-dd371928c7d6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/44904 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/43159 "Passed tests") | | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/155301 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/42787 "Passed tests") | [✅ 🛠 gtk3-gcc](https://ews-build.webkit.org/#/builders/284/builds/3989 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/49167 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/47418 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/194752 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/56197 "Built successfully") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/151944 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/40701 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/56901 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/39401 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/151644 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/46223 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/129168 "Built successfully") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/125188 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/55409 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/17786 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/54781 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/55019 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/54847 "Built successfully") | | | |

### CODE REVIEW

File: Source/WebCore/dom/Element.cpp [hunk 1]
```diff
@@ -4309,8 +4309,12 @@ void Element::focus(const FocusOptions& options)
         newTarget = findFocusDelegateForTarget(*root, options.trigger);
         if (!newTarget)
             return;
-    } else if (!isProgramaticallyFocusable(*newTarget))
+    } else if (!isProgramaticallyFocusable(*newTarget)) {
+        if (this != document->documentElement())
+            return;
```
Comment (rniwa) [inline Source/WebCore/dom/Element.cpp]: In this case, I think it's easier to read do to this:
```cpp
if (this == document->documentElement())
    document->setFocusedElement(nullptr);
return;
```

**Review by rniwa — APPROVED** [review]

File: Source/WebCore/dom/Element.cpp (hunk 1 above)
Comment (Ahmad-S792) [reply Source/WebCore/dom/Element.cpp]: Done.
